File: web/src/robotBuilder/transforms.ts
import type { Transform3 } from "../api";

export type Vec3 = [number, number, number];
export type Mat3 = [Vec3, Vec3, Vec3];
export type Mat4 = [Vec3, Vec3, Vec3, Vec3] | number[][];

const AXIS_EPS = 1e-9;
const ALIGN_EPS = 1e-8;
export const PATTERN_MATCH_TOL_IN = 0.02;

export function vec(x = 0, y = 0, z = 0): Vec3 {
  return [x, y, z];
}

export function add(a: Vec3, b: Vec3): Vec3 {
  return [a[0] + b[0], a[1] + b[1], a[2] + b[2]];
}

export function sub(a: Vec3, b: Vec3): Vec3 {
  return [a[0] - b[0], a[1] - b[1], a[2] - b[2]];
}

export function scale(a: Vec3, s: number): Vec3 {
  return [a[0] * s, a[1] * s, a[2] * s];
}

export function dot(a: Vec3, b: Vec3): number {
  return a[0] * b[0] + a[1] * b[1] + a[2] * b[2];
}

export function cross(a: Vec3, b: Vec3): Vec3 {
  return [a[1] * b[2] - a[2] * b[1], a[2] * b[0] - a[0] * b[2], a[0] * b[1] - a[1] * b[0]];
}

export function norm(a: Vec3): number {
  return Math.hypot(a[0], a[1], a[2]);
}

export function normalize(a: Vec3): Vec3 {
  const n = norm(a);
  if (n < AXIS_EPS) throw new Error("mount axis must be a non-zero vector");
  return scale(a, 1 / n);
}

export function identity4(): number[][] {
  return [
    [1, 0, 0, 0],
    [0, 1, 0, 0],
    [0, 0, 1, 0],
    [0, 0, 0, 1],
  ];
}

function mul3(a: number[][], b: Vec3): Vec3 {
  return [
    a[0][0] * b[0] + a[0][1] * b[1] + a[0][2] * b[2],
    a[1][0] * b[0] + a[1][1] * b[1] + a[1][2] * b[2],
    a[2][0] * b[0] + a[2][1] * b[1] + a[2][2] * b[2],
  ];
}

export function planeBasis(axis: Vec3): [Vec3, Vec3] {
  const normal = normalize(axis);
  const helper: Vec3 = Math.abs(normal[2]) < 0.9 ? [0, 0, 1] : [1, 0, 0];
  const u = normalize(cross(helper, normal));
  return [u, cross(normal, u)];
}

export function eulerMatrix(rollRad: number, pitchRad: number, yawRad: number): number[][] {
  const cr = Math.cos(rollRad);
  const sr = Math.sin(rollRad);
  const cp = Math.cos(pitchRad);
  const sp = Math.sin(pitchRad);
  const cy = Math.cos(yawRad);
  const sy = Math.sin(yawRad);
  return [
    [cy * cp, cy * sp * sr - sy * cr, cy * sp * cr + sy * sr],
    [sy * cp, sy * sp * sr + cy * cr, sy * sp * cr - cy * sr],
    [-sp, cp * sr, cp * cr],
  ];
}

export function matrixFromPose(pose?: Transform3 | null): number[][] {
  const row = pose || {};
  const rot = eulerMatrix(
    ((row.rollDeg || 0) * Math.PI) / 180,
    ((row.pitchDeg || 0) * Math.PI) / 180,
    ((row.yawDeg || 0) * Math.PI) / 180,
  );
  return [
    [rot[0][0], rot[0][1], rot[0][2], row.x || 0],
    [rot[1][0], rot[1][1], rot[1][2], row.y || 0],
    [rot[2][0], rot[2][1], rot[2][2], row.z || 0],
    [0, 0, 0, 1],
  ];
}

export function matrixToEuler(rotation: number[][]): [number, number, number] {
  const pitch = Math.asin(Math.min(1, Math.max(-1, -rotation[2][0])));
  const cosinePitch = Math.cos(pitch);
  if (Math.abs(cosinePitch) > 1e-8) {
    return [Math.atan2(rotation[2][1], rotation[2][2]), pitch, Math.atan2(rotation[1][0], rotation[0][0])];
  }
  return [0, pitch, Math.atan2(-rotation[0][1], rotation[1][1])];
}

export function poseFromMatrix(matrix: number[][]): Transform3 {
  const [roll, pitch, yaw] = matrixToEuler(matrix);
  return {
    x: matrix[0][3],
    y: matrix[1][3],
    z: matrix[2][3],
    rollDeg: (roll * 180) / Math.PI,
    pitchDeg: (pitch * 180) / Math.PI,
    yawDeg: (yaw * 180) / Math.PI,
  };
}

export function invertTransform(matrix: number[][]): number[][] {
  const r = [
    [matrix[0][0], matrix[1][0], matrix[2][0]],
    [matrix[0][1], matrix[1][1], matrix[2][1]],
    [matrix[0][2], matrix[1][2], matrix[2][2]],
  ];
  const t: Vec3 = [matrix[0][3], matrix[1][3], matrix[2][3]];
  const invT = scale(mul3(r, t), -1);
  return [
    [r[0][0], r[0][1], r[0][2], invT[0]],
    [r[1][0], r[1][1], r[1][2], invT[1]],
    [r[2][0], r[2][1], r[2][2], invT[2]],
    [0, 0, 0, 1],
  ];
}

export function mul4(a: number[][], b: number[][]): number[][] {
  const out = identity4();
  for (let i = 0; i < 4; i += 1) {
    for (let j = 0; j < 4; j += 1) {
      out[i][j] = a[i][0] * b[0][j] + a[i][1] * b[1][j] + a[i][2] * b[2][j] + a[i][3] * b[3][j];
    }
  }
  return out;
}

export function relativeTransform(parent: number[][], child: number[][]): number[][] {
  return mul4(invertTransform(parent), child);
}

export function transformPoint(matrix: number[][], point: Vec3): Vec3 {
  return add(mul3(matrix, point), [matrix[0][3], matrix[1][3], matrix[2][3]]);
}

export function transformDirection(matrix: number[][], direction: Vec3): Vec3 {
  return mul3(matrix, direction);
}

function skew(v: Vec3): number[][] {
  return [
    [0, -v[2], v[1]],
    [v[2], 0, -v[0]],
    [-v[1], v[0], 0],
  ];
}

function mulMat3(a: number[][], b: number[][]): number[][] {
  const out = [
    [0, 0, 0],
    [0, 0, 0],
    [0, 0, 0],
  ];
  for (let i = 0; i < 3; i += 1) {
    for (let j = 0; j < 3; j += 1) {
      out[i][j] = a[i][0] * b[0][j] + a[i][1] * b[1][j] + a[i][2] * b[2][j];
    }
  }
  return out;
}

function addMat3(a: number[][], b: number[][]): number[][] {
  return [
    [a[0][0] + b[0][0], a[0][1] + b[0][1], a[0][2] + b[0][2]],
    [a[1][0] + b[1][0], a[1][1] + b[1][1], a[1][2] + b[1][2]],
    [a[2][0] + b[2][0], a[2][1] + b[2][1], a[2][2] + b[2][2]],
  ];
}

function scaleMat3(a: number[][], s: number): number[][] {
  return [
    [a[0][0] * s, a[0][1] * s, a[0][2] * s],
    [a[1][0] * s, a[1][1] * s, a[1][2] * s],
    [a[2][0] * s, a[2][1] * s, a[2][2] * s],
  ];
}

export function axisAngle(axis: Vec3, thetaRad: number): number[][] {
  const n = normalize(axis);
  const k = skew(n);
  const kk = mulMat3(k, k);
  const eye = [
    [1, 0, 0],
    [0, 1, 0],
    [0, 0, 1],
  ];
  return addMat3(addMat3(eye, scaleMat3(k, Math.sin(thetaRad))), scaleMat3(kk, 1 - Math.cos(thetaRad)));
}

export function rotationAligning(source: Vec3, destination: Vec3): number[][] {
  const src = normalize(source);
  const dest = normalize(destination);
  const cosine = Math.min(1, Math.max(-1, dot(src, dest)));
  if (cosine > 1 - ALIGN_EPS) {
    return [
      [1, 0, 0],
      [0, 1, 0],
      [0, 0, 1],
    ];
  }
  if (cosine < -1 + ALIGN_EPS) {
    const helper: Vec3 = Math.abs(src[0]) < 0.9 ? [1, 0, 0] : [0, 1, 0];
    return axisAngle(normalize(cross(src, helper)), Math.PI);
  }
  const crossed = cross(src, dest);
  const k = skew(crossed);
  const kk = mulMat3(k, k);
  const eye = [
    [1, 0, 0],
    [0, 1, 0],
    [0, 0, 1],
  ];
  return addMat3(addMat3(eye, k), scaleMat3(kk, (1 - cosine) / dot(crossed, crossed)));
}

export function signedAngleAround(fromVec: Vec3, toVec: Vec3, axis: Vec3): number {
  const normal = normalize(axis);
  let start = sub(fromVec, scale(normal, dot(fromVec, normal)));
  let dest = sub(toVec, scale(normal, dot(toVec, normal)));
  const startN = norm(start);
  const destN = norm(dest);
  if (startN < AXIS_EPS || destN < AXIS_EPS) return 0;
  start = scale(start, 1 / startN);
  dest = scale(dest, 1 / destN);
  return Math.atan2(dot(normal, cross(start, dest)), Math.min(1, Math.max(-1, dot(start, dest))));
}

export function poseEulerRad(pose?: Transform3 | null): [number, number, number] {
  return [((pose?.pitchDeg || 0) * Math.PI) / 180, ((pose?.yawDeg || 0) * Math.PI) / 180, ((pose?.rollDeg || 0) * Math.PI) / 180];
}

File: python/talongym/robot/transforms.py
"""Rigid transforms and exact mount-to-mount snap solving."""

from __future__ import annotations

import math
from typing import Any

import numpy as np

from talongym.robot.contract import AssemblyError

_AXIS_EPS = 1e-9
_ALIGN_EPS = 1e-8
PATTERN_MATCH_TOL_IN = 0.02  # ~0.5 mm


def _normalize(vector: np.ndarray) -> np.ndarray:
    vector = np.asarray(vector, dtype=np.float64).reshape(3)
    norm = float(np.linalg.norm(vector))
    if norm < _AXIS_EPS:
        raise AssemblyError("mount axis must be a non-zero vector")
    return vector / norm


def _skew(vector: np.ndarray) -> np.ndarray:
    x, y, z = (float(vector[0]), float(vector[1]), float(vector[2]))
    return np.array([[0.0, -z, y], [z, 0.0, -x], [-y, x, 0.0]], dtype=np.float64)


def plane_basis(axis: np.ndarray) -> tuple[np.ndarray, np.ndarray]:
    normal = _normalize(axis)
    helper = np.array([0.0, 0.0, 1.0]) if abs(float(normal[2])) < 0.9 else np.array([1.0, 0.0, 0.0])
    u_axis = np.cross(helper, normal)
    u_axis = u_axis / float(np.linalg.norm(u_axis))
    v_axis = np.cross(normal, u_axis)
    return u_axis, v_axis


def euler_matrix(roll_rad: float, pitch_rad: float, yaw_rad: float) -> np.ndarray:
    """FTC Rz(yaw) * Ry(pitch) * Rx(roll), matching mjcf_field._robot_quat."""
    cr, sr = math.cos(roll_rad), math.sin(roll_rad)
    cp, sp = math.cos(pitch_rad), math.sin(pitch_rad)
    cy, sy = math.cos(yaw_rad), math.sin(yaw_rad)
    return np.array(
        [
            [cy * cp, cy * sp * sr - sy * cr, cy * sp * cr + sy * sr],
            [sy * cp, sy * sp * sr + cy * cr, sy * sp * cr - cy * sr],
            [-sp, cp * sr, cp * cr],
        ],
        dtype=np.float64,
    )


def matrix_from_pose(pose: dict[str, Any] | None) -> np.ndarray:
    row = pose or {}
    matrix = np.eye(4, dtype=np.float64)
    matrix[:3, :3] = euler_matrix(
        math.radians(float(row.get("rollDeg") or 0.0)),
        math.radians(float(row.get("pitchDeg") or 0.0)),
        math.radians(float(row.get("yawDeg") or 0.0)),
    )
    matrix[0, 3] = float(row.get("x") or 0.0)
    matrix[1, 3] = float(row.get("y") or 0.0)
    matrix[2, 3] = float(row.get("z") or 0.0)
    return matrix


def matrix_to_euler(rotation: np.ndarray) -> tuple[float, float, float]:
    pitch = math.asin(float(np.clip(-rotation[2, 0], -1.0, 1.0)))
    cosine_pitch = math.cos(pitch)
    if abs(cosine_pitch) > 1e-8:
        yaw = math.atan2(float(rotation[1, 0]), float(rotation[0, 0]))
        roll = math.atan2(float(rotation[2, 1]), float(rotation[2, 2]))
    else:
        yaw = math.atan2(float(-rotation[0, 1]), float(rotation[1, 1]))
        roll = 0.0
    return roll, pitch, yaw


def pose_from_matrix(matrix: np.ndarray) -> dict[str, float]:
    roll, pitch, yaw = matrix_to_euler(matrix[:3, :3])
    return {
        "x": float(matrix[0, 3]),
        "y": float(matrix[1, 3]),
        "z": float(matrix[2, 3]),
        "rollDeg": math.degrees(roll),
        "pitchDeg": math.degrees(pitch),
        "yawDeg": math.degrees(yaw),
    }


def invert_transform(matrix: np.ndarray) -> np.ndarray:
    rotation = matrix[:3, :3]
    inverse = np.eye(4, dtype=np.float64)
    inverse[:3, :3] = rotation.T
    inverse[:3, 3] = -rotation.T @ matrix[:3, 3]
    return inverse


def relative_transform(parent: np.ndarray, child: np.ndarray) -> np.ndarray:
    return invert_transform(parent) @ child


def transform_point(matrix: np.ndarray, point: np.ndarray) -> np.ndarray:
    return matrix[:3, :3] @ np.asarray(point, dtype=np.float64) + matrix[:3, 3]


def transform_direction(matrix: np.ndarray, direction: np.ndarray) -> np.ndarray:
    return matrix[:3, :3] @ np.asarray(direction, dtype=np.float64)


def rotation_aligning(source: np.ndarray, destination: np.ndarray) -> np.ndarray:
    source = _normalize(source)
    destination = _normalize(destination)
    cosine = float(np.clip(np.dot(source, destination), -1.0, 1.0))
    if cosine > 1.0 - _ALIGN_EPS:
        return np.eye(3, dtype=np.float64)
    if cosine < -1.0 + _ALIGN_EPS:
        helper = np.array([1.0, 0.0, 0.0]) if abs(float(source[0])) < 0.9 else np.array([0.0, 1.0, 0.0])
        axis = _normalize(np.cross(source, helper))
        return axis_angle(axis, math.pi)
    crossed = np.cross(source, destination)
    skew = _skew(crossed)
    return np.eye(3, dtype=np.float64) + skew + skew @ skew * ((1.0 - cosine) / float(np.dot(crossed, crossed)))


def axis_angle(axis: np.ndarray, theta_rad: float) -> np.ndarray:
    normal = _normalize(axis)
    skew = _skew(normal)
    return np.eye(3, dtype=np.float64) + math.sin(theta_rad) * skew + (skew @ skew) * (1.0 - math.cos(theta_rad))


def signed_angle_around(from_vec: np.ndarray, to_vec: np.ndarray, axis: np.ndarray) -> float:
    normal = _normalize(axis)
    start = from_vec - np.dot(from_vec, normal) * normal
    dest = to_vec - np.dot(to_vec, normal) * normal
    start_norm = float(np.linalg.norm(start))
    dest_norm = float(np.linalg.norm(dest))
    if start_norm < _AXIS_EPS or dest_norm < _AXIS_EPS:
        return 0.0
    start = start / start_norm
    dest = dest / dest_norm
    sine = float(np.dot(normal, np.cross(start, dest)))
    cosine = float(np.clip(np.dot(start, dest), -1.0, 1.0))
    return math.atan2(sine, cosine)


def _project_onto_plane(vector: np.ndarray, origin: np.ndarray, axis: np.ndarray) -> np.ndarray:
    offset = vector - origin
    normal = _normalize(axis)
    return offset - np.dot(offset, normal) * normal


def solve_mount_transform(
    parent_world: np.ndarray,
    parent_mount: dict[str, Any],
    child_mount: dict[str, Any],
    *,
    parent_index: tuple[int, int] | None = None,
    child_index: tuple[int, int] | None = None,
    spin_deg: float = 0.0,
    secondary: dict[str, Any] | None = None,
) -> np.ndarray:
    """Place the child so its mount hole coincides with the parent, axes anti-parallel."""
    from talongym.robot.mounts import hole_in_part, mount_axis

    parent_hole = transform_point(parent_world, hole_in_part(parent_mount, parent_index))
    parent_axis = transform_direction(parent_world, mount_axis(parent_mount))
    child_hole = hole_in_part(child_mount, child_index)
    child_axis = mount_axis(child_mount)
    desired_axis = -_normalize(parent_axis)
    rotation = rotation_aligning(child_axis, desired_axis)
    if secondary:
        spin_deg = _spin_from_secondary(
            parent_world,
            parent_hole,
            desired_axis,
            rotation,
            child_hole,
            child_mount,
            secondary,
        )
    if abs(spin_deg) > 1e-12:
        rotation = axis_angle(desired_axis, math.radians(spin_deg)) @ rotation
    translation = parent_hole - rotation @ child_hole
    matrix = np.eye(4, dtype=np.float64)
    matrix[:3, :3] = rotation
    matrix[:3, 3] = translation
    return matrix


def _spin_from_secondary(
    parent_world: np.ndarray,
    parent_hole: np.ndarray,
    desired_axis: np.ndarray,
    rotation: np.ndarray,
    child_hole: np.ndarray,
    child_mount: dict[str, Any],
    secondary: dict[str, Any],
) -> float:
    from talongym.robot.mounts import hole_in_part

    parent_second = transform_point(parent_world, hole_in_part(secondary["parent_mount"], secondary.get("parent_index")))
    child_second = hole_in_part(secondary["child_mount"], secondary.get("child_index"))
    translation = parent_hole - rotation @ child_hole
    child_second_world = rotation @ child_second + translation
    parent_radial = _project_onto_plane(parent_second, parent_hole, desired_axis)
    child_radial = _project_onto_plane(child_second_world, parent_hole, desired_axis)
    parent_radius = float(np.linalg.norm(parent_radial))
    child_radius = float(np.linalg.norm(child_radial))
    if abs(parent_radius - child_radius) > PATTERN_MATCH_TOL_IN:
        raise AssemblyError(
            f"secondary hole/pattern constraint does not match ({parent_radius:.4f} in vs {child_radius:.4f} in)"
        )
    if parent_radius < PATTERN_MATCH_TOL_IN:
        return 0.0
    return math.degrees(signed_angle_around(child_radial, parent_radial, desired_axis))
